File: Projects/rosalind/REAR.java
import java.io.*;
import java.util.*;
import java.util.stream.*;

public class REAR {
    private static int reversal(List<Integer> s, List<Integer> t) {
        final int n = s.size();
        final Map<List<Integer>, Integer> distance = new HashMap<List<Integer>, Integer>() {{ put(s, 0); }};
        final Queue<List<Integer>> queue = new LinkedList<List<Integer>>() {{ add(s); }};
        while (queue.size() > 0 && !distance.containsKey(t)) {
            List<Integer> u = new ArrayList<>(queue.remove());
            int d = distance.get(u);
            for (int i = 0; i < n; i++)
                for (int j = i+2; j <= n; j++) {
                    Collections.reverse(u.subList(i, j));
                    List<Integer> v = new ArrayList<>(u);
                    Collections.reverse(u.subList(i, j));
                    if (distance.containsKey(v)) continue;
                    distance.put(v, d+1);
                    queue.add(v);
					if (v.equals(t)) return distance.get(t);
                }
        }
        return distance.get(t);
    }
    public static void main(String[] args) throws IOException {
        Scanner in = new Scanner(System.in);
        do {
            List<Integer> s = Arrays.stream(in.nextLine().split(" ")).map(Integer::new).collect(Collectors.toList());
            List<Integer> t = Arrays.stream(in.nextLine().split(" ")).map(Integer::new).collect(Collectors.toList());
            System.out.println(reversal(s, t));
        } while (in.hasNext() && in.nextLine().equals(""));
    }
}

File: Projects/rosalind/REAR.kt
package rosalind

import java.io.IOException
import java.util.*
import java.util.stream.Collectors

object REAR {
    private fun reversal(s: List<Int>, t: List<Int>): Int {
        val n = s.size
        val distance: MutableMap<List<Int?>, Int?> = mutableMapOf(s to 0)
        val queue = LinkedList<List<Int>>()
        queue.add(s)
        while (queue.size > 0 && !distance.containsKey(t)) {
            val u: List<Int> = ArrayList(queue.remove())
            val d = distance[u]!!
            for (i in 0 until n) for (j in i + 2..n) {
                Collections.reverse(u.subList(i, j))
                val v: List<Int> = ArrayList(u)
                Collections.reverse(u.subList(i, j))
                if (distance.containsKey(v)) continue
                distance[v] = d + 1
                queue.add(v)
                if (v == t) return distance[t]!!
            }
        }
        return distance[t]!!
    }

    @Throws(IOException::class)
    @JvmStatic
    fun main(args: Array<String>) {
        val stdin = Scanner(System.`in`)
        do {
            val s = stdin.nextLine().split(" ").map(String::toInt)
            val t = stdin.nextLine().split(" ").map(String::toInt)
            println(reversal(s, t))
        } while (stdin.hasNext() && stdin.nextLine() == "")
        stdin.close()
    }
}